libsplinter/src/network/peer_manager/mod.rs

```diff
@@ -667,134 +672,239 @@ fn handle_notifications(
             peers,
             connector,
             subscribers,
+            local_identity,
         ),
         ConnectionManagerNotification::FatalConnectionError { endpoint, error } => {
             handle_fatal_connection(endpoint, error.to_string(), peers, subscribers)
         }
     }
 }
 
-fn handle_connected(
+// Allow clippy errors for too_many_arguments. The arguments are required
+// to avoid needing a lock in the PeerManager.
+#[allow(clippy::too_many_arguments)]
+fn handle_inbound_connection(
     endpoint: String,
     identity: String,
     connection_id: String,
     unreferenced_peers: &mut HashMap<String, UnreferencedPeer>,
     peers: &mut PeerMap,
     connector: Connector,
     subscribers: &mut Vec<Sender<PeerManagerNotification>>,
+    local_identity: &str,
 ) {
-    if let Some(mut peer_metadata) = peers.get_peer_from_endpoint(&endpoint).cloned() {
+    info!(
+        "Received peer connection from {} (remote endpoint: {})",
+        identity, endpoint
+    );
+
+    // If we got an inbound counnection for an existing peer, replace old connection with
+    // this new one unless we are already connected.
+    if let Some(mut peer_metadata) = peers.get_by_peer_id(&identity).cloned() {
         match peer_metadata.status {
-            PeerStatus::Connected => {
-                let notification = PeerManagerNotification::Connected {
-                    peer: peer_metadata.id.to_string(),
-                };
-                subscribers.retain(|sender| sender.send(notification.clone()).is_ok());
+            PeerStatus::Disconnected { .. } => {
+                info!(
+                    "Adding inbound connection to Disconnected peer: {}",
+                    identity
+                );
             }
             PeerStatus::Pending => {
-                if identity != peer_metadata.id {
-                    if let Err(err) = connector.remove_connection(&endpoint) {
-                        error!("Unable to clean up mismatched identity connection: {}", err);
-                    }
-
-                    // tell subscribers this Peer is currently disconnected
-                    let notification = PeerManagerNotification::Disconnected {
-                        peer: peer_metadata.id.to_string(),
-                    };
-
-                    subscribers.retain(|sender| sender.send(notification.clone()).is_ok());
-                    error!(
-                        "Peer {} (via {}) presented a mismatched identity {}",
-                        identity, endpoint, peer_metadata.id
+                info!("Adding inbound connection to pending peer: {}", identity);
```

Comment:
Be consistent with capitalization of the status describing the peer (compare with line 707)

Reply:
fixed